### PR TITLE
Improve API docs : document traits

### LINKED ISSUE #314: Improve API docs : document traits.
The API documentation for important classes like `Application` don't have traits documented automatically because the comment strings don't use the expected `#:`. They just use `#`. https://docs.enthought.com/envisage/api/envisage.html#module-envisage.application

https://github.com/enthought/envisage/blob/ae477a4e4919681006c9e84bb4f9edb77b65bba8/envisage/application.py#L50-L56

Addressing this issue has the additional reward that other projects can link to traits on envisage classes, specifically to traits defined on classes using intersphinx.

Every single class in envisage does not need to be fixed but all of the important classes that applications built using envisage need their traits documented. A non-exhaustive list of such classes is 

- `Application` and `IApplication`
- `TasksApplication`, `TaskFactory`, `TaskExtension`
- `Plugin`, `CorePlugin` and `IPlugin`
- `PluginManager` and `IPluginManager`
- `ServiceOffer`

### PR DESCRIPTION
**Pull request by aaronayres35**
fixes #314

This PR simply replaces "#" comments on trait declarations with "#:" so that the traits are automatically documented.  It does so to all the classes listed in the issue (done in first commit)

It also does so to a few other envisage classes.  Namely some other HasTraits classes in api modules, as well as a few classes I saw imported in an application built using envisage.
Some of these may be overkill, and there also may be classes I missed.

### CODE REVIEW

File: envisage/application.py [hunk 1]
```diff
@@ -47,64 +47,64 @@ class Application(HasTraits):
...
     plugin_manager = Instance(IPluginManager)
 
-    # The service registry.
+    #: The service registry.
     service_registry = Instance(IServiceRegistry)
 
     #### Private interface ####################################################
 
-    # The import manager.
+    #: The import manager.
```
Comment (rahulporuri) [inline envisage/application.py:107]: We probably don't want to document the private traits.

```suggestion
    # The import manager.
```

File: envisage/extension_registry.py [hunk 2]
```diff
@@ -69,23 +69,23 @@ class ExtensionRegistry(HasTraits):
     # Protected 'ExtensionRegistry' interface.
     ###########################################################################
 
-    # A dictionary of extensions, keyed by extension point.
+    #: A dictionary of extensions, keyed by extension point.
```
Comment (rahulporuri) [inline envisage/extension_registry.py:72]:
```suggestion
    # A dictionary of extensions, keyed by extension point.
```

File: envisage/extension_registry.py [hunk 3]
```diff
@@ -69,23 +69,23 @@ class ExtensionRegistry(HasTraits):
...
     _extensions = Dict
 
-    # The extension points that have been added *explicitly*.
+    #: The extension points that have been added *explicitly*.
```
Comment (rahulporuri) [inline envisage/extension_registry.py:75]:
```suggestion
    # The extension points that have been added *explicitly*.
```

File: envisage/extension_registry.py [hunk 4]
```diff
@@ -69,23 +69,23 @@ class ExtensionRegistry(HasTraits):
...
+    #:
+    #: These are called when extensions are added to or removed from an
+    #: extension point.
+    #:
+    #: e.g. Dict(extension_point, [weakref.ref(callable)])
+    #:
+    #: A listener is any Python callable with the following signature:-
+    #:
+    #: def listener(extension_registry, extension_point_changed_event):
+    #:     ...
```
Comment (rahulporuri) [inline envisage/extension_registry.py:88]:
```suggestion
    # Extension listeners.
    #
    # These are called when extensions are added to or removed from an
    # extension point.
    #
    # e.g. Dict(extension_point, [weakref.ref(callable)])
    #
    # A listener is any Python callable with the following signature:-
    #
    # def listener(extension_registry, extension_point_changed_event):
    #     ...
```

File: envisage/plugin.py [hunk 5]
```diff
@@ -45,47 +45,47 @@ class Plugin(ExtensionProvider):
...
     #### 'IServiceUser' interface #############################################
 
-    # The service registry that the object's services are stored in.
+    #: The service registry that the object's services are stored in.
     service_registry = Property(Instance(IServiceRegistry))
 
     #### Private interface ####################################################
 
-    # The Ids of the services that were automatically registered.
+    #: The Ids of the services that were automatically registered.
```
Comment (rahulporuri) [inline envisage/plugin.py:88]:
```suggestion
    # The Ids of the services that were automatically registered.
```

File: envisage/plugin_manager.py [hunk 6]
```diff
@@ -190,7 +190,7 @@ def stop_plugin(self, plugin=None, plugin_id=None):
 
     #### Protected 'PluginManager' ############################################
 
-    # The plugins that the manager manages!
+    #: The plugins that the manager manages!
```
Comment (rahulporuri) [inline envisage/plugin_manager.py:193]:
```suggestion
    # The plugins that the manager manages!
```

File: envisage/provider_extension_registry.py [hunk 7]
```diff
@@ -32,7 +32,7 @@ class ProviderExtensionRegistry(ExtensionRegistry):
 
     #### Protected 'ProviderExtensionRegistry' interface ######################
 
-    # The extension providers that populate the registry.
+    #: The extension providers that populate the registry.
```
Comment (rahulporuri) [inline envisage/provider_extension_registry.py:35]:
```suggestion
    # The extension providers that populate the registry.
```

File: envisage/service_registry.py [hunk 8]
```diff
@@ -35,31 +35,31 @@ class ServiceRegistry(HasTraits):
...
+    #: 'obj' is the object that is registered (any old, Python object!).
+    #:
+    #: 'properties' is the arbitrary dictionary of properties that were
+    #: registered with the object.
     _services = Dict
 
-    # The next service Id (service Ids are never persisted between process
-    # invocations so this is simply an ever increasing integer!).
+    #: The next service Id (service Ids are never persisted between process
+    #: invocations so this is simply an ever increasing integer!).
```
Comment (rahulporuri) [inline envisage/service_registry.py:62]:
```suggestion
    # The next service Id (service Ids are never persisted between process
    # invocations so this is simply an ever increasing integer!).
```

File: envisage/service_registry.py [hunk 9]
```diff
@@ -35,31 +35,31 @@ class ServiceRegistry(HasTraits):
...
+    #:
+    #: where:
+    #:
+    #: 'protocol_name' is the (possible dotted) name of the interface, type or
+    #: class that the object is registered against.
+    #:
+    #: 'obj' is the object that is registered (any old, Python object!).
+    #:
+    #: 'properties' is the arbitrary dictionary of properties that were
+    #: registered with the object.
```
Comment (rahulporuri) [inline envisage/service_registry.py:58]:
```suggestion
    # The services in the registry.
    #
    # { service_id : (protocol_name, obj, properties) }
    #
    # where:
    #
    # 'protocol_name' is the (possible dotted) name of the interface, type or
    # class that the object is registered against.
    #
    # 'obj' is the object that is registered (any old, Python object!).
    #
    # 'properties' is the arbitrary dictionary of properties that were
    # registered with the object.
```

File: envisage/ui/tasks/tasks_application.py [hunk 10]
```diff
@@ -44,111 +44,113 @@ class TasksApplication(Application):
...
     #### Protected interface ##################################################
 
-    # An 'explicit' exit is when the the 'exit' method is called.
-    # An 'implicit' exit is when the user closes the last open window.
+    #: An 'explicit' exit is when the the 'exit' method is called.
+    #: An 'implicit' exit is when the user closes the last open window.
     _explicit_exit = Bool(False)
 
-    # Application state.
+    #: Application state.
```
Comment (rahulporuri) [inline envisage/ui/tasks/tasks_application.py:153]:
```suggestion
    # Application state.
```

File: envisage/ui/tasks/tasks_application.py [hunk 11]
```diff
@@ -44,111 +44,113 @@ class TasksApplication(Application):
...
     window_closed = Event(
         Instance("envisage.ui.tasks.task_window_event.TaskWindowEvent")
     )
 
     #### Protected interface ##################################################
 
-    # An 'explicit' exit is when the the 'exit' method is called.
-    # An 'implicit' exit is when the user closes the last open window.
+    #: An 'explicit' exit is when the the 'exit' method is called.
+    #: An 'implicit' exit is when the user closes the last open window.
```
Comment (rahulporuri) [inline envisage/ui/tasks/tasks_application.py:150]:
```suggestion
    # An 'explicit' exit is when the the 'exit' method is called.
    # An 'implicit' exit is when the user closes the last open window.
```

**Review by rahulporuri — CHANGES_REQUESTED** [review]

**Review by rahulporuri — APPROVED** [review]
LGTM. I haven't tried building and checking the API docs to see if the traits are now documented on the classes.

**Comment by aaronayres35** [comment]
> LGTM. I haven't tried building and checking the API docs to see if the traits are now documented on the classes. Have you checked this aaron?

I did not check every class but did a random sample of a couple classes and in those I checked the traits were showing up

File: envisage/application.py [hunk 12]
```diff
@@ -47,59 +47,59 @@ class Application(HasTraits):
...
     #### 'Application' interface ##############################################
 
-    # These traits allow application developers to build completely different
-    # styles of extensible application. It allows Envisage to be used as a
-    # framework for frameworks ;^)
-    #
-    # The extension registry.
+    #: These traits allow application developers to build completely different
+    #: styles of extensible application. It allows Envisage to be used as a
+    #: framework for frameworks ;^)
```
Comment (aaronayres35) [inline envisage/application.py:94]: These comments are specific to all the traits below, this carries them into the docs for `extension_registry` specifically.  Removing now

**Pull request merged**